### ISSUE TITLE
Join not producing output

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#4800]
Repro is with Zed commit 40a2d27 and files from the [ZNG format zed-sample-data](https://github.com/brimdata/zed-sample-data/tree/main/zng).

I was brainstorming on some `join` use cases, and thought it'd be cool to collect up Zeek `weird` logs and join them by `uid` to their associated `conn` records. I therefore expected this to work, but it produces no output.

```
$ zq -version
Version: v1.10.0-6-g40a2d275

$ cat weird_join.zed 
file conn.zng.gz
| inner join (
  file weird.zng.gz | collect(this) by uid
) on uid=uid weirdos:=this

$ zq -I weird_join.zed 
[no output]
```

In trying to figure out what's wrong, I confirmed there's at least one matching pair of join keys in the respective data sources and extract those subsets of the data to separate ZSON-based inputs.

```
$ zq 'count() by uid | sort -r count | head 1' conn.zng.gz weird.zng.gz
{uid:"Ctr6GV2SVX0UaxrMP9",count:5(uint64)}

$ zq -z 'uid=="Ctr6GV2SVX0UaxrMP9"' conn.zng.gz | tee conn.zson
{_path:"conn",ts:2018-03-24T17:16:05.065763Z,uid:"Ctr6GV2SVX0UaxrMP9",id:{orig_h:10.47.5.155,orig_p:40728(port=uint16),resp_h:91.189.91.23,resp_p:80(port)},proto:"tcp"(=zenum),service:null(string),duration:10.011295s,orig_bytes:21396(uint64),resp_bytes:98972170(uint64),conn_state:"SF",local_orig:null(bool),local_resp:null(bool),missed_bytes:0(uint64),history:"SAdtDTttttTfFa",orig_pkts:74548(uint64),orig_ip_bytes:4351284(uint64),resp_pkts:136942(uint64),resp_ip_bytes:205392620(uint64),tunnel_parents:null(|[string]|)}

$ zq -z 'uid=="Ctr6GV2SVX0UaxrMP9"' weird.zng.gz | tee weird.zson
{_path:"weird",ts:2018-03-24T17:16:06.322585Z,uid:"Ctr6GV2SVX0UaxrMP9",id:{orig_h:10.47.5.155,orig_p:40728(port=uint16),resp_h:91.189.91.23,resp_p:80(port)},name:"possible_split_routing",addl:null(string),notice:false,peer:"zeek"}
{_path:"weird",ts:2018-03-24T17:16:06.323092Z,uid:"Ctr6GV2SVX0UaxrMP9",id:{orig_h:10.47.5.155,orig_p:40728(port=uint16),resp_h:91.189.91.23,resp_p:80(port)},name:"TCP_ack_underflow_or_misorder",addl:null(string),notice:false,peer:"zeek"}
{_path:"weird",ts:2018-03-24T17:16:06.323882Z,uid:"Ctr6GV2SVX0UaxrMP9",id:{orig_h:10.47.5.155,orig_p:40728(port=uint16),resp_h:91.189.91.23,resp_p:80(port)},name:"data_before_established",addl:null(string),notice:false,peer:"zeek"}
{_path:"weird",ts:2018-03-24T17:16:06.326478Z,uid:"Ctr6GV2SVX0UaxrMP9",id:{orig_h:10.47.5.155,orig_p:40728(port=uint16),resp_h:91.189.91.23,resp_p:80(port)},name:"above_hole_data_without_any_acks",addl:null(string),notice:false,peer:"zeek"}
```

Running the same join logic on those smaller ZSON inputs, now I get output.

```
$ cat weird_join_mini.zed
file conn.zson
| inner join (
  file weird.zson | collect(this) by uid
) on uid=uid weirdos:=this

$ zq -Z -I weird_join_mini.zed 
{
    _path: "conn",
    ts: 2018-03-24T17:16:05.065763Z,
    uid: "Ctr6GV2SVX0UaxrMP9",
    id: {
        orig_h: 10.47.5.155,
        orig_p: 40728 (port=uint16),
        resp_h: 91.189.91.23,
        resp_p: 80 (port)
    },
    proto: "tcp" (=zenum),
    service: null (string),
    duration: 10.011295s,
    orig_bytes: 21396 (uint64),
    resp_bytes: 98972170 (uint64),
    conn_state: "SF",
    local_orig: null (bool),
    local_resp: null (bool),
    missed_bytes: 0 (uint64),
    history: "SAdtDTttttTfFa",
    orig_pkts: 74548 (uint64),
    orig_ip_bytes: 4351284 (uint64),
    resp_pkts: 136942 (uint64),
    resp_ip_bytes: 205392620 (uint64),
    tunnel_parents: null (|[string]|),
    weirdos: {
        uid: "Ctr6GV2SVX0UaxrMP9",
        collect: [
            {
                _path: "weird",
                ts: 2018-03-24T17:16:06.322585Z,
                uid: "Ctr6GV2SVX0UaxrMP9",
                id: {
                    orig_h: 10.47.5.155,
                    orig_p: 40728 (port),
                    resp_h: 91.189.91.23,
                    resp_p: 80 (port)
                },
                name: "possible_split_routing",
                addl: null (string),
                notice: false,
                peer: "zeek"
            },
            {
                _path: "weird",
                ts: 2018-03-24T17:16:06.323092Z,
                uid: "Ctr6GV2SVX0UaxrMP9",
                id: {
                    orig_h: 10.47.5.155,
                    orig_p: 40728 (port),
                    resp_h: 91.189.91.23,
                    resp_p: 80 (port)
                },
                name: "TCP_ack_underflow_or_misorder",
                addl: null (string),
                notice: false,
                peer: "zeek"
            },
            {
                _path: "weird",
                ts: 2018-03-24T17:16:06.323882Z,
                uid: "Ctr6GV2SVX0UaxrMP9",
                id: {
                    orig_h: 10.47.5.155,
                    orig_p: 40728 (port),
                    resp_h: 91.189.91.23,
                    resp_p: 80 (port)
                },
                name: "data_before_established",
                addl: null (string),
                notice: false,
                peer: "zeek"
            },
            {
                _path: "weird",
                ts: 2018-03-24T17:16:06.326478Z,
                uid: "Ctr6GV2SVX0UaxrMP9",
                id: {
                    orig_h: 10.47.5.155,
                    orig_p: 40728 (port),
                    resp_h: 91.189.91.23,
                    resp_p: 80 (port)
                },
                name: "above_hole_data_without_any_acks",
                addl: null (string),
                notice: false,
                peer: "zeek"
            }
        ]
    }
}
```

In conclusion, I can't see a reason why there was output when using the excerpt inputs but no output with the original full data sources.

**Comment by philrz** [comment]
Verified in Zed commit da25095.

The Zed program from the original repro now produces the expected output. For instance, here's the first timestamp-sorted output from the `join` that contains some reference to HTTP.

```
$ zq -version
Version: v1.10.0-8-gda250959

$ zq -Z -I weird_join.zed '| sort ts | HTTP | head 1'
{
    _path: "conn",
    ts: 2018-03-24T17:15:21.006035Z,
    uid: "C8B46DDUQM0UjAS3j",
    id: {
        orig_h: 10.164.94.120,
        orig_p: 33749 (port=uint16),
        resp_h: 10.47.3.200,
        resp_p: 8000 (port)
    },
    proto: "tcp" (=zenum),
    service: null (string),
    duration: 4.163948s,
    orig_bytes: 9 (uint64),
    resp_bytes: 0 (uint64),
    conn_state: "RSTO",
    local_orig: null (bool),
    local_resp: null (bool),
    missed_bytes: 0 (uint64),
    history: "ShAaDTR",
    orig_pkts: 8 (uint64),
    orig_ip_bytes: 450 (uint64),
    resp_pkts: 6 (uint64),
    resp_ip_bytes: 328 (uint64),
    tunnel_parents: null (|[string]|),
    weirdos: {
        uid: "C8B46DDUQM0UjAS3j",
        collect: [
            {
                _path: "weird",
                ts: 2018-03-24T17:15:21.007184Z,
                uid: "C8B46DDUQM0UjAS3j",
                id: {
                    orig_h: 10.164.94.120,
                    orig_p: 33749 (port),
                    resp_h: 10.47.3.200,
                    resp_p: 8000 (port)
                },
                name: "bad_HTTP_request",
                addl: null (string),
                notice: false,
                peer: "zeek"
            }
        ]
    }
}
```

Thanks @mattnibs!
